src/klein/_plating.py

```diff
@@ -25,17 +34,74 @@ def _should_return_json(request):
     return bool(request.args.get(b"json"))
 
 
-def json_serialize(item):
+@inlineCallbacks
+def resolveDeferredObjects(root):
     """
-    A function similar to L{dumps}.
+    Wait on possibly nested L{Deferred}s that represent a JSON
+    serializable object.
+
+    @param root: A JSON-serializable object that may contain
+        L{Deferred}s, or a Deferred that will resolve to a
+        JSON-serializable object
+
+    @return: A L{Deferred} that fires with a L{Deferred}-free version
+        of C{root}, or that fails with the first exception
+        encountered.
     """
-    def helper(unknown):
-        if isinstance(unknown, PlatedElement):
-            return unknown._asJSON()
+
+    result = [None]
+    setResult = partial(setitem, result, 0)
+    stack = [(root, setResult)]
+
+    while stack:
+        mightBeDeferred, setter = stack.pop()
+        obj = yield mightBeDeferred
+        if isinstance(obj, ATOM_TYPES):
+            setter(obj)
+        elif isinstance(obj, list):
+            parent = [None] * len(obj)
+            setter(parent)
+            stack.extend(
+                reversed([
+                    (child, partial(setitem, parent, i))
+                    for i, child in enumerate(obj)
+                ])
+            )
+        elif isinstance(obj, tuple):
+            parent = [None] * len(obj)
+            setter(tuple(parent))
+
+            def setTupleItem(i, value, parent=parent, setter=setter):
+                parent[i] = value
+                setter(tuple(parent))
+
+            stack.extend(
+                reversed([
+                    (child, partial(setTupleItem, i))
+                    for i, child in enumerate(obj)
+                ])
+            )
+        elif isinstance(obj, dict):
+            parent = {}
+            setter(parent)
+            for key, value in reversed(list(obj.items())):
+                pair = [None, None]
+                setKey = partial(setitem, pair, 0)
+
+                def setValue(value, pair=pair, parent=parent):
```

Comment:
You could take the function definition out of the loop, and use partial here -- I think that looks nicer than the default-args-tricks, and it's not like you're not using partial all over anyway...